node/src/tasks/msp_charge_fees.rs

```diff
@@ -0,0 +1,129 @@
+use anyhow::anyhow;
+use sc_tracing::tracing::*;
+use shc_actors_framework::event_bus::EventHandler;
+use shc_blockchain_service::commands::BlockchainServiceInterface;
+use shc_blockchain_service::events::NotifyPeriod;
+use shc_blockchain_service::types::Tip;
+use shc_common::types::MaxUsersToCharge;
+use shc_common::types::StorageProviderId;
+use sp_core::Get;
+use storage_hub_runtime::Balance;
+
+use crate::tasks::{FileStorageT, MspForestStorageHandlerT, StorageHubHandler};
+
+const LOG_TARGET: &str = "msp-charge-fees-task";
+const MIN_DEBT: Balance = 0;
+
+pub struct MspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    storage_hub_handler: StorageHubHandler<FL, FSH>,
+}
+
+impl<FL, FSH> Clone for MspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    fn clone(&self) -> MspChargeFeesTask<FL, FSH> {
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+        }
+    }
+}
+
+impl<FL, FSH> MspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    pub fn new(storage_hub_handler: StorageHubHandler<FL, FSH>) -> Self {
+        Self {
+            storage_hub_handler,
+        }
+    }
+}
+
+/// Handles the [`NotifyPeriod`] event.
+///
+/// This event is triggered every X amount of blocks.
+///
+/// This task will:
+/// - Charge users for the MSP when triggered
+impl<FL, FSH> EventHandler<NotifyPeriod> for MspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    async fn handle_event(&mut self, event: NotifyPeriod) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Charging users",
+        );
+
+        let own_provider_id = self
+            .storage_hub_handler
+            .blockchain
+            .query_storage_provider_id(None)
+            .await?;
```

Comment:
Shouldn't MSP task have access to their own msp id ?

Reply:
It would be nicer to not have to do this, indeed.

But when starting up, the only known thing is the account - the provider id associated is known by the runtime.